street_network/src/transform/separate_cycletracks.rs

```diff
@@ -0,0 +1,215 @@
+
+use crate::{osm, BufferType, Direction, LaneSpec, LaneType, OriginalRoad, StreetNetwork};
+
+/// Find cycleway segments that exist as separate objects, parallel to a main road. Merge (or
+/// "snap") them into the main road, inserting a buffer lane to represent the physical division.
+pub fn snap_cycleways(streets: &mut StreetNetwork) {
+    for cycleway in find_cycleways(streets) {
+        cycleway.debug(streets);
+        snap(streets, cycleway);
+    }
+}
+
+// We're only pattern matching on one type of separate cycleway right now. This represents a single
+// RawRoad that's parallel to one or more main_roads.
+//
+// X--X
+// C  M
+// C  M
+// C  M
+// C  X
+// C  M
+// C  M
+// X--X
+//
+// C is the cycleway segment. X are intersections. M are main roads -- note there are two matching
+// up to this one cycleway. The '-'s are short connector roads between the two.
+struct Cycleway {
+    cycleway: OriginalRoad,
+    cycleway_center: PolyLine,
+    // Just to distinguish different cycleways when debugging
+    debug_idx: usize,
+    main_road_i1: osm::NodeID,
+    main_road_i2: osm::NodeID,
+    main_roads: Vec<OriginalRoad>,
+}
+
+impl Cycleway {
+    fn debug(&self, streets: &StreetNetwork) {
+        streets.debug_road(self.cycleway, format!("cycleway {}", self.debug_idx));
+        streets.debug_intersection(self.main_road_i1, format!("i1 of {}", self.debug_idx));
+        streets.debug_intersection(self.main_road_i2, format!("i2 of {}", self.debug_idx));
+        for x in &self.main_roads {
+            streets.debug_road(*x, format!("main road along {}", self.debug_idx));
+        }
+    }
+}
+
+fn find_cycleways(streets: &StreetNetwork) -> Vec<Cycleway> {
+    const SHORT_ROAD_THRESHOLD: Distance = Distance::const_meters(10.0);
+
+    let mut cycleways = Vec::new();
+    for (cycleway_id, cycleway_road) in &streets.roads {
+        if cycleway_road.is_cycleway() {
+            // Look at other roads connected to both endpoints. One of them should be "very short."
+            let mut main_road_endpoints = Vec::new();
+            for i in [cycleway_id.i1, cycleway_id.i2] {
+                let mut candidates = Vec::new();
+                for r in streets.roads_per_intersection(i) {
+                    let road = &streets.roads[&r];
+                    if road.length() < SHORT_ROAD_THRESHOLD {
+                        candidates.push(r);
+                    }
+                }
+                if candidates.len() == 1 {
+                    main_road_endpoints.push(candidates[0].other_side(i));
+                }
+            }
+
+            if main_road_endpoints.len() == 2 {
+                // Often the main road parallel to this cycleway segment is just one road, but it
+                // might be more.
+                let main_road_i1 = main_road_endpoints[0];
+                let main_road_i2 = main_road_endpoints[1];
+                // Find all main road segments "parallel to" this cycleway, by pathfinding between
```

Comment:
We could perhaps be more paranoid here by checking that the cycleway center line and all of the main road center lines are nice and parallel. If they curve away from each other a bit, then even if the connection roads are under our threshold of 10 meters, we may lose some geometric information by snapping.